Fix KleidiAI compilation errors with -DGGML_NATIVE=OFF (issue #14464)

baonudesifeizhai opened the pull request:
This commit fixes compilation errors that occur when building with -DGGML_NATIVE=OFF, which resulted in zero-size arrays in KleidiAI code.

Changes made:

1. kernels.cpp:
   - Add conditional compilation around gemm_gemv_kernels array
   - Provide fallback empty array when no ARM features available
   - Guard kernel selection functions with feature checks

2. kleidiai.cpp:
   - Replace GGML_ASSERT(kernels) with null pointer checks
   - Return appropriate error codes when no kernels available
   - Prevent crashes when KleidiAI is unavailable

3. CMakeLists.txt:
   - Add architecture check to only enable KleidiAI on ARM systems
   - Fix KleidiAI download URL (GitHub -> GitLab)
   - Use git clone instead of archive download for reliability

Fixes: https://github.com/ggml-org/llama.cpp/issues/14464
Tested: Successfully compiles with -DGGML_NATIVE=OFF on x86_64

*Make sure to read the [contributing guidelines](https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md) before submitting a PR*

<img width="542" height="680" alt="1752607464722" src="https://github.com/user-attachments/assets/9eee5f04-8b63-4066-8c53-10af68f5017f" />
<img width="510" height="449" alt="1752607473355" src="https://github.com/user-attachments/assets/5ab80822-ca0a-4976-8944-be8bc4b7bb9a" />

baonudesifeizhai committed:
Fix KleidiAI compilation errors with -DGGML_NATIVE=OFF (issue #14464)

This commit fixes compilation errors that occur when building with
-DGGML_NATIVE=OFF, which resulted in zero-size arrays in KleidiAI code.

Changes made:

1. kernels.cpp:
   - Add conditional compilation around gemm_gemv_kernels array
   - Provide fallback empty array when no ARM features available
   - Guard kernel selection functions with feature checks

2. kleidiai.cpp:
   - Replace GGML_ASSERT(kernels) with null pointer checks
   - Return appropriate error codes when no kernels available
   - Prevent crashes when KleidiAI is unavailable

3. CMakeLists.txt:
   - Add architecture check to only enable KleidiAI on ARM systems
   - Fix KleidiAI download URL (GitHub -> GitLab)
   - Use git clone instead of archive download for reliability

Fixes: https://github.com/ggml-org/llama.cpp/issues/14464
Tested: Successfully compiles with -DGGML_NATIVE=OFF on x86_64

baonudesifeizhai committed:
Fix KleidiAI compilation with improved fallback and error handling

- Revert to GitHub tarball download for stability
- Add debug logging for KleidiAI kernel fallback scenarios
- Improve error messages when no suitable kernels available
- Keep ARM64 architecture requirement for KleidiAI enabling
- Ensure graceful fallback to standard CPU implementation

chaxu01 commented on ggml/src/ggml-cpu/kleidiai/kleidiai.cpp:
```diff
@@ -344,7 +350,9 @@ class tensor_traits : public ggml::cpu::tensor_traits {
-        GGML_ASSERT(ctx.kernels);
+        if (!ctx.kernels) {
+            return -1;  // No suitable kernel available
+        }
```
Consider keeping the GGML_ASSERT(ctx.kernels) to catch misconfigurations early, or at least add a GGML_LOG_DEBUG message to warn when no suitable kernel is available.

ggerganov replied:
I don't understand the change - if `repack()` returns -1, we will immediately assert on line 405:

```c++
    auto OK            = tensor_traits->repack(tensor, data, size);

    GGML_ASSERT(OK == 0);
```

baonudesifeizhai committed:
Address PR review feedback: Keep GGML_ASSERT for early misconfiguration detection

- Add GGML_LOG_DEBUG message to warn when no suitable kernel is available
- Keep GGML_ASSERT(ctx.kernels) to catch misconfigurations early as suggested by chaxu01
- This provides both debugging information and early error detection

baonudesifeizhai committed:
Address PR review feedback: Improve repack() error handling while keeping GGML_ASSERTs in compute functions

CISC commented:
@baonudesifeizhai It looks like this PR is ready, but needs rebasing?

CISC commented:
> > @baonudesifeizhai It looks like this PR is ready, but needs rebasing?
> 
> already rebase .... there comes so many changes ...

Such is the life in the fast lane. :)

You seem to have made a bad merge though.
